text/001x-battery_improvements.md

```diff
@@ -0,0 +1,293 @@
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The proposed message is:
+
+```xml
+    <message id="369" name="BATTERY_STATUS_V2">
+      <description>Battery dynamic information.
+        This should be streamed (nominally at 1Hz).
+        Static/invariant battery information is sent in SMART_BATTERY_INFO.
+        Note that smart batteries should set the MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL bit to indicate that supplied capacity values are relative to a battery that is known to be full.
+        Power monitors would not set this bit, indicating that capacity_consumed is relative to drone power-on, and that other values are estimated based on the assumption that the battery was full on power-on.
+      </description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the whole battery pack (not internal electronics). INT16_MAX field not provided.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="UINT32_MAX">Battery voltage (total). UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="current" units="mA" invalid="UINT32_MAX">Battery current (through all cells/loads). UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="capacity_consumed" units="mAh" invalid="UINT32_MAX">Consumed charge. UINT32_MAX: field not provided. This is either the consumption since power-on or since the battery was full, depending on the value of MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL.</field>
+      <field type="uint32_t" name="capacity_remaining" units="mAh" invalid="UINT32_MAX">Remaining charge (until empty). UINT32_MAX: field not provided. Note: If MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL is unset, this value is based on the assumption the battery was full when the system was powered.</field>
+      <field type="uint8_t" name="percent_remaining" units="%" invalid="UINT8_MAX">Remaining battery energy. Values: [0-100], UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="status_flags" display="bitmask" enum="MAV_BATTERY_STATUS_FLAGS">Fault, health, readiness, and other status indications.</field>
+    </message>
+```
+
+`MAV_BATTERY_STATUS_FLAGS` is a new enum that includes faults, charge state, and battery mode:
+
+```xml
+    <enum name="MAV_BATTERY_STATUS_FLAGS" bitmask="true">
+      <description>Battery status flags for fault, health and state indication.</description>
+      <entry value="1" name="MAV_BATTERY_STATUS_FLAGS_READY_TO_USE">
```

Comment:
@dakejahl I might be overthinking this, but should this perhaps be "not ready to use" ? i.e it we're constructing it from all the other faults so we're not saying it is ready to use, but that it is "not unready to use". 

As it is now, you'd have to set the bit even if you didn't know the readiness, and by setting it you're saying its good to go. Whereas if we reverse it we're just saying - some flag that indicates there is a problem is set.

Anyway, not certain on this. I like a readiness flag too - just not sure we can be ready = not unready

Reply:
@hendjoshsr71 First, thanks for the review. Looks like a lot more issues than I first thought

This is supposed to reflect "there are no known issues that would prevent this flying" so how about MAV_BATTERY_STATUS_FLAGS_NOT_READY_TO_USE ? 

The actual issues that set this would depend on the person setting up the system. If it is OK to charge in place, then this would not be a fault that set this.

Reply:
>@dakejahl I might be overthinking this

I think you might :laughing: 

Hmmm I'm not sure... You are saying you want it to be "not ready to use" since this would then correlate with a flag that gives us the reason for it being not ready? I would say you would infer this in the same way as seeing "ready to use" be set to zero. The other part is that an empty message (all zeroes) would be default to "not ready" which is a nice side effect.

Or maybe I don't see the point.

>There is a question for you and @dakejahl on inverting the "ready to use flag" to "unready to use". The former is what users want, but the later is what we can realistically and reliably provide.

I don't understand how we can know it's "not ready" but can't know it's "ready"? They're the same thing just inverse.

Reply:
@dakejahl What we do is check a bunch of things which mean it is **not ready** to use. So what we are saying is there are no flags that indicate it is not ready to use. Then we're inverting that and saying, "if there are no flags that say it is not ready to use it is ready to use."

Are they semantically the same? Technically they are not, because we only detect "known issues". It could be that there is an unknown issue we do not know about - so we're telling people it is ready to use when it is not. A legal expert would care, but perhaps we do not need to. 

> The other part is that an empty message (all zeroes) would be default to "not ready" which is a nice side effect.

It means that if you don't have the ability to report or detect faults you have to set this as ready all the time.

That's kind of my point - you're not actually saying its ready, you're just saying you don't know of any faults.

Reply:
Ahh so you're saying "the system is not not ready" does not mean the same thing as "the system is ready"...? Essentially saying we could just not know the state. I honestly don't have a strong opinion on it, I think semantically you are correct but I sure do hate double negatives lol. So yeah we can change it to "not ready" and then the check would be if it's unset, easy enough. I'll leave that up to you.